### ISSUE TITLE
problems with MS:1002676 (Protein-pair level global FDR) and MS:1002677 (Residue pair-level global FDR)

### ISSUE THREAD

**Issue by colin-combe** [HUPO-PSI/psi-ms-CV#170]
### Describe the question or discussion

Hi,
in short MS:1002676 (Protein-pair level global FDR) and MS:1002677 (Residue pair-level global FDR) have the wrong value type. 

They are strings described by a regular expression (see pg.22 of 1.2.0 mzId spec) but their value type is given as a double. This is also related to #150, you can see the problem in this image:
 
![Screenshot from 2022-12-01 10-38-28](https://user-images.githubusercontent.com/977607/205024992-f2a2ced9-d94a-46df-96cc-05e4c1c009bc.png)

However, there are a couple of other things:

- our understanding of the mzIdentML spec is that CV terms giving scores can generally be used either to score a PSM or Peptide group, or to give the cut-off threshold for that score by including it in the Threshold element. Therefore, we think the regular expression for these terms (MS:1002665) should be changed to allow the possibility that it is just a number (for when MS:1002676 or MS:1002677 are placed in a Threshold element to give the cut-off score).
- we think the '+' or '-' sign in the score should be optional after the 'e' when using scientific notation (related #142) 

So, in additional to changing the value type of MS:1002676 and MS:1002677 to a string, we propose changing regex MS:1002665 from:
`([0-9]+)[.]([a|b]):([0-9]+|null):([\-+]?[0-9]+(?:[.][0-9]+)?(?:[Ee][+-][0-9]+)?):(true|false)`
to
`(([0-9]+)[.]([a|b]):([0-9]+|null):([-+]?[0-9]+(?:[.][0-9]+)?(?:[Ee][+-]?[0-9]+)?):(true|false))|([-+]?[0-9]+(?:[.][0-9]+)?(?:[Ee][+-]?[0-9]+)?)`
(have i got that right @lutzfischer )

feedback appreciated,
cheers,
Colin

**Comment by lutzfischer** [comment]
Some `-` should be escaped:
`(([0-9]+)[.]([a|b]):([0-9]+|null):([-+]?[0-9]+(?:[.][0-9]+)?(?:[Ee][+\-]?[0-9]+)?):(true|false))|([\-+]?[0-9]+(?:[.][0-9]+)?(?:[Ee][+\-]?[0-9]+)?)`

or seemingly have `[-+]` instead of `[+\-]`:
`(([0-9]+)[.]([a|b]):([0-9]+|null):([-+]?[0-9]+(?:[.][0-9]+)?(?:[Ee][-+]?[0-9]+)?):(true|false))|([-+]?[0-9]+(?:[.][0-9]+)?(?:[Ee][-+]?[0-9]+)?)`

**Comment by mobiusklein** [comment]
Can you please name the capture groups in the new regular expression using `?<name-here>` to make it clearer what the value  they are matching is? This new pattern appears to OR the original pattern with a new construct which I can't tell the intent of.

**Comment by colin-combe** [comment]
Hi Josh - thanks for the feedback. 

I named the capturing groups using the syntax you suggested:
`((?<int_ID>[0-9]+)[.](?<SUFFIX>[a|b]):(?<POS>[0-9]+|null):(?<SCORE_OR_VALUE>[-+]?[0-9]+(?:[.][0-9]+)?(?:[Ee][-+]?[0-9]+)?):(?<PASS_THRESHOLD>true|false))|(?<SCORE_ONLY>[-+]?[0-9]+(?:[.][0-9]+)?(?:[Ee][-+]?[0-9]+)?)`

see https://regex101.com/r/Ivg7Mj/1
see also p.22 of the mzIdentML 1.2.0 spec doc.
I notice the regex is case sensitive for the 'true'/'false' part but spec doc gives it uppercase and lowercase in different places. We should make the whole thing case insensitive.
Also, the naming of the groups has made it incompatible with some flavours of regex.

>  This new pattern appears to OR the original pattern with a new construct which I can't tell the intent of.

Indeed, this is exactly what it does. 

The new construct is meant to be for when this CV term is just used to record the cut-off threshold applied and is placed inside a Threshold element. We thought this was the way things are meant to work. This is how _some_ other scores appear to work. Do you know anything about this?

In short - i think the question is 'what should be the CV mapping rules for the Threshold element' (https://github.com/HUPO-PSI/mzIdentML/issues/131). I am thinking the only CV mapping rule should be "one or more child of '[statistical threshold](https://www.ebi.ac.uk/ols/ontologies/ms/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FMS_1002482&lang=en&viewMode=All&siblings=false)'".

Thoughts?

**Comment by colin-combe** [comment]
> I am thinking the only CV mapping rule [for Threshold] should be "one or more child of '[statistical threshold](https://www.ebi.ac.uk/ols/ontologies/ms/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FMS_1002482&lang=en&viewMode=All&siblings=false)'".

preceding comment is wrong because not all thresholds are statistical thresholds (i believe)

>> This new pattern appears to OR the original pattern with a new construct which I can't tell the intent of.

> The new construct is meant to be for when this CV term is just used to record the cut-off threshold applied and is placed inside a Threshold element. 

yep, we think this is what we should do -  OR the original pattern with a new construct that allows just a number, for use when these CV terms are used to state what the applied threshold was.

**Comment by colin-combe** [comment]
@lutzfischer - what about case sensitivity of 'true' and 'false'? 

Fig 4 of the 1.2.0 spec doc (p.22) shows them in both upper and lower case in different parts of the fig?

**Comment by mobiusklein** [comment]
I think it's better to fix the figure, not complicate parsing even further. The [XSD specification](https://www.w3.org/TR/xmlschema-2/#boolean) dictates `true` and `false` are legal, and all other casings of the words are not.
